File: Programmers/지형 이동.cpp
#include<bits/stdc++.h>
#define f(i,l,r) for(int i=l;i<r;++i)
using namespace std;

vector<vector<int>> l;
struct info{
	int diff,i,j;
	bool operator<(const info &oth)const{
		return diff>oth.diff;
	}
};
priority_queue<info> q;
int n,h,di[]={0,0,1,-1},dj[]={1,-1,0,0};
bool vis[300][300];

void dfs(int i, int j){
	vis[i][j]=true;
	f(d,0,4){
		int ni=i+di[d],nj=j+dj[d];
		if(ni>=0&&nj>=0&&ni<n&&nj<n&&!vis[ni][nj]){
			int diff=abs(l[i][j]-l[ni][nj]);
			if(diff>h)q.push({diff,ni,nj});
			else dfs(ni,nj);
		}
	}
}

int solution(vector<vector<int>> land, int height) {
	land.swap(l);
	n=l.size();
	h=height;
	q.push({0,0,0});
	int ans=0;
	while(!q.empty()){
		int diff=q.top().diff,i=q.top().i,j=q.top().j;q.pop();
		if(vis[i][j])continue;
		ans+=diff;
		dfs(i,j);
	}
    return ans;
}

int main(){
	ios_base::sync_with_stdio(0); cin.tie(0); cout.tie(0);
	cout<<solution({{10, 11, 10, 11}, {2, 21, 20, 10}, {1, 20, 21, 11}, {2, 1, 2, 1}}, 1);
	return 0;
}

File: Programmers/지형 이동.swift
import Foundation
typealias Info = (Int,Int,Int)
var heap=[(-1,-1,-1),(0,0,0)]

func push_heap(_ info:Info){
	var i=heap.count
	heap.append(info)
	while i>1{
		let j=i>>1
		if heap[j].0>heap[i].0{
			heap.swapAt(j, i)
		}else{
			break
		}
		i=j
	}
}

func pop_heap() -> Info{
	heap.swapAt(1, heap.count-1)
	let ret=heap.removeLast()
	var i=1
	while true{
		var j=i<<1,idx=0
		if j<heap.count,heap[i].0>heap[j].0{
			idx=j
		}
		if j+1<heap.count,heap[i].0>heap[j+1].0,heap[j].0>heap[j+1].0{
			idx=j+1
		}
		if idx<1{
			break
		}
		heap.swapAt(i, idx)
		i=idx
	}
	return ret
}

func solution(_ land:[[Int]], _ height:Int) -> Int {
	let di=[0,0,1,-1],dj=[1,-1,0,0],n=land.count
	var vis=Array(repeating: Array(repeating: false, count: n), count:n)
	
	func dfs(_ i:Int, _ j:Int){
		vis[i][j]=true
		for d in 0...3{
			let ni=i+di[d],nj=j+dj[d]
			if ni>=0,nj>=0,ni<n,nj<n,!vis[ni][nj]{
				let diff=abs(land[i][j]-land[ni][nj])
				if diff>height{
					push_heap((diff,ni,nj))
				}else{
					dfs(ni, nj)
				}
			}
		}
	}
	
	var ans=0
	while(heap.count>1){
		let info=pop_heap()
		if vis[info.1][info.2]{continue}
		ans+=info.0
		dfs(info.1, info.2)
	}
    return ans
}
